Topgrade upgrade from 16.0.4 to 16.1.2 on MacOS ARM

<!--
Thanks for taking the time to fill out this bug report!
Please make sure to
[search for existing issues](https://github.com/topgrade-rs/topgrade/issues)
before filing a new one!

Questions labeled with `Optional` can be skipped.
-->

<!--
If you're here to report about a "No asset found" error, please make sure that
an hour has been passed since the last release was made.
-->

## Erroneous Behavior

<details>
<pre>Topgrade not upgrading to 16.1.2 from 16.0.4 on MacOS ARM
</pre>
</details>


## Expected Behavior

<details>
<pre>
Expected Topgrade to upgrade from 16.0.4 to 16.1.2 without error
</pre>
</details>


## Steps to reproduce

<details>
<pre>
Running topgrade with self-update enabled causing topgrade to upgrade fail on topgrade upgrade from cc linking fail
</pre>
</details>

## Possible Cause (Optional)

<details>
<pre>
cc not targeting topgrade binary correctly while building
</pre>
</details>

## Problem persists without calling from topgrade
<!--
Execute the erroneous command directly to see if the problem persists
-->
- [x ] Yes
- [ ] No

## Did you run topgrade through `Remote Execution`

- [x ] Yes
- [ ] No

If yes, does the issue still occur when you run topgrade directly in your
remote host

- [ x] Yes
- [ ] No

## Configuration file (Optional)
<!--
Paste your configuration file inside the code block if you think this issue is
related to configuration.
-->

```toml

```

## Additional Details
- Operation System/Version

<details>
<pre>
MacOS Tahoe 26.0.1 ARM on Mac Mini M2
</pre>
</details>

- Installation

<details>
<pre>
Crates.io
</pre>
</details>

- Topgrade version (`topgrade -V`) 
<details>
<pre>
16.0.4
</pre>
</details>

## Verbose Output (`topgrade -v`)
<!--
Paste the verbose output into the pre-tags
-->

<details>
<pre>
DEBUG Current system locale is en-US
DEBUG Configuration directory /Users/nick/.config/ does not exist
DEBUG Version: 16.0.4
DEBUG OS: aarch64-apple-darwin
DEBUG Args { inner: ["topgrade", "-v"] }
DEBUG Binary path: Ok("/Users/nick/.cargo/bin/topgrade")
DEBUG self-update Feature Enabled: false
DEBUG Configuration: Config { opt: CommandLineArgs { edit_config: false, show_config_reference: false, run_in_tmux: false, cleanup: false, dry_run: false, no_retry: false, disable: [], only: [], custom_commands: [], env: [], verbose: true, keep_at_end: false, skip_notify: false, yes: None, disable_predefined_git_repos: false, config: None, remote_host_limit: None, show_skipped: false, log_filter: "warn", gen_completion: None, gen_manpage: false, no_self_update: false }, config_file: ConfigFile { include: None, misc: None, pre_commands: None, post_commands: None, commands: None, python: None, composer: None, brew: None, linux: None, git: None, containers: None, windows: None, npm: None, yarn: None, deno: None, vim: None, firmware: None, vagrant: None, flatpak: None, distrobox: None, lensfun: None, julia: None, zigup: None, vscode: None }, allowed_steps: [AM, AndroidStudio, AppMan, Aqua, Asdf, Atom, Audit, AutoCpufreq, Bin, Bob, BrewCask, BrewFormula, Bun, BunPackages, Cargo, Certbot, Chezmoi, Chocolatey, Choosenim, CinnamonSpices, ClamAvDb, Composer, Conda, ConfigUpdate, Containers, CustomCommands, DebGet, Deno, Distrobox, DkpPacman, Dotnet, Elan, Emacs, Firmware, Flatpak, Flutter, Fossil, Gcloud, Gem, Ghcup, GitRepos, GithubCliExtensions, GnomeShellExtensions, Go, Guix, Haxelib, Helix, Helm, HomeManager, JetbrainsAqua, JetbrainsClion, JetbrainsDatagrip, JetbrainsDataspell, JetbrainsGateway, JetbrainsGoland, JetbrainsIdea, JetbrainsMps, JetbrainsPhpstorm, JetbrainsPycharm, JetbrainsRider, JetbrainsRubymine, JetbrainsRustrover, JetbrainsToolbox, JetbrainsWebstorm, Jetpack, Julia, Juliaup, Kakoune, Krew, Lensfun, Lure, Macports, Mamba, Mas, Maza, Micro, MicrosoftStore, Miktex, Mise, Myrepos, Nix, Node, Opam, Pacdef, Pacstall, Pearl, Pip3, PipReview, PipReviewLocal, Pipupgrade, Pipx, Pipxu, Pixi, Pkg, Pkgin, PlatformioCore, Pnpm, Poetry, Powershell, Protonup, Pyenv, Raco, Rcm, Remotes, Restarts, Rtcl, RubyGems, Rustup, Rye, Scoop, Sdkman, SelfUpdate, Sheldon, Shell, Snap, Sparkle, Spicetify, Stack, Stew, System, Tldr, Tlmgr, Tmux, Toolbx, Uv, Vagrant, Vcpkg, Vim, VoltaPackages, Vscode, Vscodium, Waydroid, Winget, Wsl, WslUpdate, Xcodes, Yadm, Yarn, Yazi, Zigup, Zvm] }
DEBUG Cannot find "pwsh"
DEBUG Cannot find "powershell"
DEBUG Path "/Users/nick/.config/emacs" doesn't exist
DEBUG Path "/Users/nick/.emacs.d" doesn't exist
DEBUG Cannot find "doas"
DEBUG Detected "/usr/bin/sudo" as "sudo"
DEBUG Cannot find "gsudo"
DEBUG Step "Brew (ARM)"
DEBUG Detected "/opt/homebrew/bin/brew" as "/opt/homebrew/bin/brew"

── 01:36:34 - Brew ─────────────────────────────────────────────────────────────
DEBUG Executing command `/opt/homebrew/bin/brew update`
==> Updating Homebrew...
Already up-to-date.
DEBUG Executing command `/opt/homebrew/bin/brew upgrade --formula`
DEBUG Step "Brew (Intel)"
DEBUG Step "Brew"
DEBUG Detected "/opt/homebrew/bin/brew" as "brew"
DEBUG Step "Brew Cask (ARM)"
DEBUG Detected "/opt/homebrew/bin/brew" as "/opt/homebrew/bin/brew"

── 01:36:36 - Brew - Cask ──────────────────────────────────────────────────────
DEBUG Executing command `/opt/homebrew/bin/brew --repository buo/cask-upgrade`
DEBUG Executing command `/opt/homebrew/bin/brew upgrade --cask`
DEBUG Step "Brew Cask (Intel)"
DEBUG Step "Brew Cask"
DEBUG Detected "/opt/homebrew/bin/brew" as "brew"
DEBUG Step "MacPorts"
DEBUG Step "Xcodes"
DEBUG Step "Sparkle"
DEBUG Step "App Store"
DEBUG Step "System upgrade"

── 01:36:37 - macOS system update ──────────────────────────────────────────────
Finding available software
DEBUG Executing command `softwareupdate --list`
DEBUG Utf8Output { status: ExitStatus(unix_wait_status(0)), stdout: "Software Update Tool\n\nFinding available software\n", stderr: "No new software available.\n" }
No new software available.
DEBUG Step "yadm"
DEBUG Step "nix"
DEBUG Step "nix upgrade-nix"
DEBUG Step "guix"
DEBUG Step "home-manager"
DEBUG Step "asdf"
DEBUG Step "mise"
DEBUG Step "pkgin"
DEBUG Step "bun-packages"
DEBUG Step "zr"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "antibody"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "antidote"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "antigen"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "zgenom"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "zplug"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "zinit"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "zi"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "zim"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Executing command `zsh -c '[[ -n ${ZIM_HOME} ]] && print -n ${ZIM_HOME}'`
DEBUG Command failed: Err(
   0: Command failed: `zsh -c '[[ -n ${ZIM_HOME} ]] && print -n ${ZIM_HOME}'`
   1: `zsh` failed: exit status: 1 with 

Location:
   /Users/nick/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/topgrade-16.0.4/src/steps/zsh.rs:147)
DEBUG Step "oh-my-zsh"
DEBUG Detected "/bin/zsh" as "zsh"
DEBUG Step "oh-my-bash"
DEBUG Detected "/bin/bash" as "bash"
DEBUG Step "fisher"
DEBUG Step "bash-it"
DEBUG Step "oh-my-fish"
DEBUG Step "fish-plug"
DEBUG Step "fundle"
DEBUG Step "tmux"
DEBUG Step "TLDR"
DEBUG Step "pearl"
DEBUG Step "pyenv"
DEBUG Step "SDKMAN!"
DEBUG Detected "/bin/bash" as "bash"
DEBUG Step "rcm"
DEBUG Step "maza"
DEBUG Step "apm"
DEBUG Step "fossil"
DEBUG Step "elan"
DEBUG Step "rye"
DEBUG Step "rustup"
DEBUG Detected "/Users/nick/.cargo/bin/rustup" as "rustup"

── 01:36:41 - rustup ───────────────────────────────────────────────────────────
DEBUG Executing command `/Users/nick/.cargo/bin/rustup update`
info: syncing channel updates for 'stable-aarch64-apple-darwin'
info: checking for self-update

  stable-aarch64-apple-darwin unchanged - rustc 1.91.0 (f8297e351 2025-10-28)

info: cleaning up downloads & tmp directories
DEBUG Step "juliaup"
DEBUG Step ".NET"
DEBUG Step "choosenim"
DEBUG Step "cargo"
DEBUG Path "/Users/nick/.cargo" exists
DEBUG Detected "/Users/nick/.cargo/bin/cargo" as "cargo"
DEBUG Path "/Users/nick/.cargo/.crates.toml" exists

── 01:36:42 - Cargo ────────────────────────────────────────────────────────────
DEBUG Detected "/Users/nick/.cargo/bin/cargo-install-update" as "cargo-install-update"
DEBUG Executing command `/Users/nick/.cargo/bin/cargo-install-update install-update --git --all`
    Polling registry 'https://index.crates.io/'...

Package       Installed  Latest   Needs update
topgrade      v16.0.4    v16.1.2  Yes
cargo-cache   v0.8.3     v0.8.3   No
cargo-update  v18.0.0    v18.0.0  No

Updating topgrade
    Updating crates.io index
  Installing topgrade v16.1.2
    Updating crates.io index
     Locking 304 packages to latest compatible versions
      Adding console v0.15.11 (available: v0.16.1)
      Adding etcetera v0.8.0 (available: v0.11.0)
      Adding merge v0.1.0 (available: v0.2.0)
      Adding nix v0.29.0 (available: v0.30.1)
      Adding regex v1.10.6 (available: v1.12.2)
      Adding strum v0.26.3 (available: v0.27.2)
      Adding tempfile v3.10.1 (available: v3.23.0)
      Adding thiserror v1.0.69 (available: v2.0.17)
      Adding tokio v1.47.2 (available: v1.48.0)
      Adding toml v0.8.23 (available: v0.9.8)
      Adding which v6.0.3 (available: v8.0.0)
   Compiling proc-macro2 v1.0.103
   Compiling unicode-ident v1.0.22
   Compiling quote v1.0.41
   Compiling serde_core v1.0.228
   Compiling memchr v2.7.6
   Compiling libc v0.2.177
   Compiling serde v1.0.228
   Compiling crossbeam-utils v0.8.21
   Compiling regex-syntax v0.8.8
   Compiling log v0.4.28
   Compiling aho-corasick v1.1.4
   Compiling itoa v1.0.15
   Compiling bitflags v2.10.0
   Compiling pin-project-lite v0.2.16
   Compiling once_cell v1.21.3
   Compiling regex-automata v0.4.13
   Compiling syn v2.0.108
   Compiling cfg-if v1.0.4
   Compiling objc2 v0.6.3
   Compiling serde_json v1.0.145
   Compiling objc2-encode v4.1.0
   Compiling bstr v1.12.1
   Compiling version_check v0.9.5
   Compiling crossbeam-epoch v0.9.18
   Compiling crunchy v0.2.4
   Compiling powerfmt v0.2.0
   Compiling same-file v1.0.6
   Compiling deranged v0.5.5
   Compiling walkdir v2.5.0
   Compiling crossbeam-deque v0.8.6
   Compiling num-conv v0.1.0
   Compiling time-core v0.1.6
   Compiling tiny-keccak v2.0.2
   Compiling hashbrown v0.16.0
   Compiling lazy_static v1.5.0
   Compiling equivalent v1.0.2
   Compiling time v0.3.44
   Compiling objc2-core-foundation v0.3.2
   Compiling indexmap v2.12.0
   Compiling tracing-core v0.1.34
   Compiling bitflags v1.3.2
   Compiling futures-core v0.3.31
   Compiling heck v0.5.0
   Compiling utf8parse v0.2.2
   Compiling getrandom v0.2.16
   Compiling anstyle-parse v0.2.7
   Compiling proc-macro-error-attr v1.0.4
   Compiling toml_write v0.1.2
   Compiling globset v0.4.18
   Compiling regex v1.10.6
   Compiling winnow v0.7.13
   Compiling serde_derive v1.0.228
   Compiling ignore v0.4.25
   Compiling tracing-attributes v0.1.30
   Compiling autocfg v1.5.0
   Compiling find-msvc-tools v0.1.4
   Compiling globwalk v0.8.1
   Compiling anstyle-query v1.1.4
   Compiling futures-io v0.3.31
   Compiling ryu v1.0.20
   Compiling smallvec v1.15.1
   Compiling shlex v1.3.0
   Compiling syn v1.0.109
   Compiling anstyle v1.0.13
   Compiling is_terminal_polyfill v1.70.2
   Compiling colorchoice v1.0.4
   Compiling anstream v0.6.21
   Compiling cc v1.2.44
   Compiling tracing v0.1.41
   Compiling num-traits v0.2.19
   Compiling const-random-macro v0.1.16
   Compiling tracing-log v0.2.0
   Compiling matchers v0.2.0
   Compiling sharded-slab v0.1.7
   Compiling proc-macro-error v1.0.4
   Compiling thread_local v1.1.9
   Compiling either v1.15.0
   Compiling clap_lex v0.7.6
   Compiling rustix v0.38.44
   Compiling arc-swap v1.7.1
   Compiling object v0.37.3
   Compiling nu-ansi-term v0.50.3
   Compiling rustversion v1.0.22
   Compiling strsim v0.11.1
   Compiling stable_deref_trait v1.2.1
   Compiling unsafe-libyaml v0.2.11
   Compiling futures-sink v0.3.31
   Compiling option-ext v0.2.0
   Compiling owo-colors v4.2.3
   Compiling futures-channel v0.3.31
   Compiling serde_spanned v0.6.9
   Compiling toml_datetime v0.6.11
   Compiling triomphe v0.1.15
   Compiling dirs-sys v0.5.0
   Compiling toml_edit v0.22.27
   Compiling clap_builder v4.5.51
   Compiling serde_yaml v0.9.34+deprecated
   Compiling tracing-subscriber v0.3.20
   Compiling itertools v0.11.0
   Compiling mac-notification-sys v0.6.8
   Compiling toml v0.8.23
   Compiling const-random v0.1.18
   Compiling clap_derive v4.5.49
   Compiling futures-macro v0.3.31
   Compiling block2 v0.6.2
   Compiling errno v0.3.14
   Compiling gimli v0.32.3
   Compiling cfg_aliases v0.2.1
   Compiling adler2 v2.0.1
   Compiling color-spantrace v0.3.0
   Compiling eyre v0.6.12
   Compiling fastrand v2.3.0
   Compiling pin-utils v0.1.0
   Compiling futures-task v0.3.31
   Compiling base62 v2.2.3
   Compiling normpath v1.5.0
   Compiling slab v0.4.11
   Compiling siphasher v1.0.1
   Compiling futures-util v0.3.31
   Compiling rust-i18n-support v3.1.5
   Compiling clap v4.5.51
   Compiling miniz_oxide v0.8.9
   Compiling addr2line v0.25.1
   Compiling nix v0.29.0
   Compiling objc2-foundation v0.3.2
   Compiling dlv-list v0.5.2
   Compiling tracing-error v0.2.1
   Compiling dirs v6.0.0
   Compiling rust-i18n v3.1.5
   Compiling objc2-io-kit v0.3.2
   Compiling core-foundation-sys v0.8.7
   Compiling thiserror v1.0.69
   Compiling rustc-demangle v0.1.26
   Compiling home v0.5.12
   Compiling indenter v0.3.4
   Compiling parking v2.2.1
   Compiling glob v0.3.3
   Compiling hashbrown v0.14.5
   Compiling futures-lite v2.6.1
   Compiling rust-i18n-macro v3.1.5
   Compiling ordered-multimap v0.7.3
   Compiling sysinfo v0.36.1
   Compiling iana-time-zone v0.1.64
   Compiling strum_macros v0.26.4
   Compiling futures-executor v0.3.31
   Compiling merge_derive v0.1.0
   Compiling thiserror-impl v1.0.69
   Compiling signal-hook-registry v1.4.6
   Compiling mio v1.1.0
   Compiling backtrace v0.3.76
   Compiling roff v0.2.2
   Compiling unicode-width v0.2.2
   Compiling bytes v1.10.1
   Compiling topgrade v16.1.2
   Compiling console v0.15.11
   Compiling color-eyre v0.6.5
   Compiling clap_mangen v0.2.31
   Compiling tokio v1.47.2
   Compiling strum v0.26.3
   Compiling merge v0.1.0
   Compiling futures v0.3.31
   Compiling jetbrains-toolbox-updater v5.4.6
   Compiling chrono v0.4.42
   Compiling rust-ini v0.21.3
   Compiling etcetera v0.8.0
   Compiling which v6.0.3
   Compiling shellexpand v3.1.1
   Compiling tempfile v3.10.1
   Compiling clap_complete v4.5.60
   Compiling regex-split v0.1.0
   Compiling shell-words v1.1.0
   Compiling sys-locale v0.3.2
   Compiling wildmatch v2.5.0
   Compiling semver v1.0.27
   Compiling notify-rust v4.11.7
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "<1 object files omitted>" "/var/folders/31/4zkfm2312m76_lvy86j94rsh0000gn/T/rustcCAlaFm/libmac_notification_sys-f40777ab68885596.rlib" "<sysroot>/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-*.rlib" "-framework" "CoreFoundation" "-framework" "Foundation" "-lSystem" "-framework" "CoreFoundation" "-lobjc" "-framework" "IOKit" "-framework" "IOKit" "-framework" "CoreFoundation" "-lobjc" "-framework" "Foundation" "-liconv" "-lSystem" "-lc" "-lm" "-arch" "arm64" "-mmacosx-version-min=11.0.0" "-L" "/var/folders/31/4zkfm2312m76_lvy86j94rsh0000gn/T/cargo-installv819of/release/build/mac-notification-sys-2ca20317557d5459/out" "-o" "/var/folders/31/4zkfm2312m76_lvy86j94rsh0000gn/T/cargo-installv819of/release/deps/topgrade-1241d411aa772f86" "-Wl,-dead_strip" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: ld: warning: ignoring duplicate libraries: '-lSystem', '-lobjc'
          Undefined symbols for architecture arm64:
            "_LSCopyApplicationURLsForBundleIdentifier", referenced from:
                _setApplication in libmac_notification_sys-f40777ab68885596.rlib[2](54532c69c7c8f551-notify.o)
            "_OBJC_CLASS_$_NSImage", referenced from:
                 in libmac_notification_sys-f40777ab68885596.rlib[2](54532c69c7c8f551-notify.o)
          ld: symbol(s) not found for architecture arm64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: could not compile `topgrade` (bin "topgrade") due to 1 previous error
error: failed to compile `topgrade v16.1.2`, intermediate artifacts can be found at `/var/folders/31/4zkfm2312m76_lvy86j94rsh0000gn/T/cargo-installv819of`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.


Updated 0 packages.
Failed to update topgrade.

    Checking 0 git packages

Package  Installed  Latest  Needs update

No git packages need updating.
Overall updated 0 packages.
Overall failed to update 1 package: topgrade.
DEBUG Command failed: Err(
   0: Command failed: `/Users/nick/.cargo/bin/cargo-install-update install-update --git --all`
   1: `/Users/nick/.cargo/bin/cargo-install-update` failed: exit status: 101

Location:
   /Users/nick/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/topgrade-16.0.4/src/steps/generic.rs:72)
DEBUG Step "cargo" failed: 
   0: Command failed: `/Users/nick/.cargo/bin/cargo-install-update install-update --git --all`
   1: `/Users/nick/.cargo/bin/cargo-install-update` failed: exit status: 101

Location:
   /Users/nick/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/topgrade-16.0.4/src/steps/generic.rs:72
cargo failed: 
   0: Command failed: `/Users/nick/.cargo/bin/cargo-install-update install-update --git --all`
   1: `/Users/nick/.cargo/bin/cargo-install-update` failed: exit status: 101

Location:
   /Users/nick/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/topgrade-16.0.4/src/steps/generic.rs:72
Retry? (y)es/(N)o/(s)hell/(q)uit%
</pre>
</details>

FFR:
```
error: linking with cc failed: exit status: 1

|

= note:  "cc" "<1 object files omitted>" "/var/folders/31/4zkfm2312m76_lvy86j94rsh0000gn/T/rustcCAlaFm/libmac_notification_sys-f40777ab68885596.rlib" "/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-*.rlib" "-framework" "CoreFoundation" "-framework" "Foundation" "-lSystem" "-framework" "CoreFoundation" "-lobjc" "-framework" "IOKit" "-framework" "IOKit" "-framework" "CoreFoundation" "-lobjc" "-framework" "Foundation" "-liconv" "-lSystem" "-lc" "-lm" "-arch" "arm64" "-mmacosx-version-min=11.0.0" "-L" "/var/folders/31/4zkfm2312m76_lvy86j94rsh0000gn/T/cargo-installv819of/release/build/mac-notification-sys-2ca20317557d5459/out" "-o" "/var/folders/31/4zkfm2312m76_lvy86j94rsh0000gn/T/cargo-installv819of/release/deps/topgrade-1241d411aa772f86" "-Wl,-dead_strip" "-nodefaultlibs"

= note: some arguments are omitted. use --verbose to show all linker arguments

= note: ld: warning: ignoring duplicate libraries: '-lSystem', '-lobjc'

Undefined symbols for architecture arm64:

"_LSCopyApplicationURLsForBundleIdentifier", referenced from:

_setApplication in libmac_notification_sys-f40777ab68885596.rlib[2](https://github.com/topgrade-rs/topgrade/issues/54532c69c7c8f551-notify.o)

"OBJC_CLASS$_NSImage", referenced from:

in libmac_notification_sys-f40777ab68885596.rlib[2](https://github.com/topgrade-rs/topgrade/issues/54532c69c7c8f551-notify.o)

ld: symbol(s) not found for architecture arm64

clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
The problem lies with `mac-notification-sys`

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -92,9 +92,6 @@ self_update_crate = { version = "~0.42", default-features = false, optional = tr
 windows = { version = "~0.62", features = ["Win32_System_Console"] }
 windows-registry = "~0.6"
 
-[target.'cfg(macos)'.dependencies]
-mac-notification-sys = "=0.6.2"
-
 [profile.release]
 lto = true
 

